### PR TITLE
Use levenshtein distance to find alt packages

### LINKED ISSUE #19: Make `users` more helpful when package name not found
Current behaviour (in v0.3)
```julia
julia> users("ParameterJUMP")
ERROR: ParameterJUMP not in General

Stacktrace:
 [1] _get_pkg_uuid(pkg_name::String, registry_name::String; depots::Vector{String}, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ PkgDeps ~/.julia/packages/PkgDeps/6nJkw/src/PkgDeps.jl:66
 [2] _get_pkg_uuid
   @ ~/.julia/packages/PkgDeps/6nJkw/src/PkgDeps.jl:61 [inlined]
 [3] #users#8
   @ ~/.julia/packages/PkgDeps/6nJkw/src/PkgDeps.jl:187 [inlined]
 [4] users(pkg_name::String)
   @ PkgDeps ~/.julia/packages/PkgDeps/6nJkw/src/PkgDeps.jl:187
 [5] top-level scope
   @ REPL[10]:1
```

This is both a bit more scary and a bit less helpful than it could be. Especially when it comes to a package name typo; the package is called `ParameterJuMP` not `ParameterJUMP` (also this is a real example i ran into when trying using this package in practice for the first time today).

I wonder if
- the message could be improved to
  - be clearer on what went wrong
  - suggest similarly named packages 
  - direct users to the `pkg_registry_name` keyword 
- the function shouldn't error, but should instead show a message (e.g. as an `@info` or `@warn`), then return `nothing`

Example: 
```julia
julia> users("ParameterJUMP")
PackageNotFoundError: No package with the name "ParameterJUMP" was found in the "General" registry.
Perhaps you meant: "ParameterJuMP"
Or you can search in another registry using `users("ParameterJUMP"; pkg_registry_name="OtherRegistry")`
```
or
```julia
julia> users("ParameterJUMP")
┌ Info: The package "ParameterJUMP" was not found in the "General" registry.
│ Perhaps you meant: "ParameterJuMP"
└ Or you can search in another registry using `users("ParameterJUMP"; pkg_registry_name="OtherRegistry")`
```

### PR DESCRIPTION
**Pull request by mattBrzezinski**
## Resolves: #19 

Since this is a pre-1.0 package, I have no issues with not having a super-clean interface for somethings in exchange for the ability to more quickly add features and iterate over the package. When this package matures, cleaning this up will be more highly prioritized, but as this is currently just an off-hand utility package.

- Moved utility functions into `/src/utilities.jl`
- If there is a failure to find the UUID of a package by its string, output alternative packages which are closely named
- All `reachable_registry()` calls return an Array of instances, this simplifies things a lot more
- Fixed bug with `users()` would not allow you to look for usages of a package outside of `pkg_registry`. For example, if I wanted to look up all usages of `DataFrames` in an internal repository, I could not unless `DataFrames` was registered in the internal repository.
- Set the compat to `Julia1.6`

### CODE REVIEW

File: src/utilities.jl [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    return [pkg for (distance, pkg) in pkg_distances if distance <= max_distance]
+end
+
+
+"""
+Get the package name from a UUID
+"""
+
+function _get_pkg_name(uuid::UUID, registries=Array{RegistryInstance})
```
Comment (nickrobinson251) [inline src/utilities.jl]: typo here? this is a type, but i suspect it's meant to be an empty array?
```suggestion
function _get_pkg_name(uuid::UUID, registries=RegistryInstance[])
```

File: src/utilities.jl [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+"""
+Get the UUID from a package name and the registry it is in.
+Specify a registry name as well to avoid ambiguity with same package names in multiple registries.
+"""
+function _get_pkg_uuid(
+    pkg_name::String, registry_name::String;
+    depots::Union{String, Vector{String}}=Base.DEPOT_PATH,
+)
+    registry = reachable_registries(registry_name; depots=depots)
+    return _get_pkg_uuid(pkg_name, first(registry))
```
Comment (nickrobinson251) [inline src/utilities.jl]: seems easier to follow the code if we use `only` here (also i think if somehow `reachable_registries(registry_name; depots=depots)` returns multiple registries, then i see no reason to prefer the first, but also that shouldn't be possible afaiu)
```suggestion
    registry = only(reachable_registries(registry_name; depots=depots))
    return _get_pkg_uuid(pkg_name, registry)
```

File: src/utilities.jl [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    else
+        alt_packages = _find_alternative_packages(pkg_name, collect(keys(registry.pkgs)))
+
+        warning = "The package $(pkg_name) was not found in the $(registry.name) registry."
+
+        if !isempty(alt_packages)
+            warning *= "\nPerhaps you meant: $(string(alt_packages))"
+        end
+
+        warning *= "\nOr you can search in another registry using `users(\"$(pkg_name)\"; pkg_registry_name=\"OtherRegistry\")`"
```
Comment (nickrobinson251) [inline src/utilities.jl]: we can simplify this
```suggestion
        warning *= "\nOr you can search in another registry using `users(\"$(pkg_name)\", \"OtherRegistry\")`"
```

File: test/runtests.jl [hunk 4]
```diff
@@ -3,20 +3,20 @@ using Test
 using UUIDs
 
 const DEPOT = joinpath(@__DIR__, "resources")
-const FOOBAR_REGISTRY = reachable_registries("Foobar"; depots=DEPOT)
+const FOOBAR_REGISTRY = first(reachable_registries("Foobar"; depots=DEPOT))
```
Comment (nickrobinson251) [inline test/runtests.jl]:
```suggestion
const FOOBAR_REGISTRY = only(reachable_registries("Foobar"; depots=DEPOT))
```

File: test/runtests.jl [hunk 5]
```diff
@@ -48,7 +48,7 @@ end
 
 @testset "reachable_registries" begin
     @testset "specfic registry -- $(typeof(v))" for v in ("Foobar", ["Foobar"])
-        registry = reachable_registries("Foobar"; depots=DEPOT)
+        registry = first(reachable_registries("Foobar"; depots=DEPOT))
```
Comment (nickrobinson251) [inline test/runtests.jl]:
```suggestion
        registry = only(reachable_registries("Foobar"; depots=DEPOT))
```

File: src/PkgDeps.jl [hunk 6]
```diff
@@ -173,10 +123,11 @@ end
...
     users(pkg_name::String, pkg_registry::RegistryInstance; kwargs...)
 
-Find the users of the package named `pkg_name` which is registered in `pkg_registry`.
+Find the users of `pkg_name` which is registered in `pkg_registry_name`.
+Optionally limit search for users of `pkg_name` in specific registries by passing in kwarg `registries`.
 """
-function users(pkg_name::String, pkg_registry::RegistryInstance; kwargs...)
-    uuid = _get_pkg_uuid(pkg_name, pkg_registry)
+function users(pkg_name::String, pkg_registry_name::RegistryInstance; kwargs...)
+    uuid = _get_pkg_uuid(pkg_name, pkg_registry_name)
```
Comment (nickrobinson251) [inline src/PkgDeps.jl]: why change the variable name here? this isn't just the registry_name, this is the registry.
```suggestion
function users(pkg_name::String, pkg_registry::RegistryInstance; kwargs...)
    uuid = _get_pkg_uuid(pkg_name, pkg_registry)
```

File: src/PkgDeps.jl [hunk 7]
```diff
@@ -173,10 +123,11 @@ end
 """
     users(pkg_name::String, pkg_registry::RegistryInstance; kwargs...)
 
-Find the users of the package named `pkg_name` which is registered in `pkg_registry`.
+Find the users of `pkg_name` which is registered in `pkg_registry_name`.
+Optionally limit search for users of `pkg_name` in specific registries by passing in kwarg `registries`.
```
Comment (nickrobinson251) [inline src/PkgDeps.jl]: The `registries` is already documented (since docstrings for the same method get appended together), so no need to document this in two places :)
```suggestion
Find the users of `pkg_name` which is registered in `pkg_registry`.
```

File: src/utilities.jl [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        end
+    end
+
+    throw(NoUUIDMatch("No package found with the UUID $uuid"))
+end
+
+function _get_pkg_name(uuid::UUID; kwargs...)
+    registries = reachable_registries(; kwargs...)
+    return _get_pkg_name(uuid, registries)
+end
```
Comment (nickrobinson251) [inline src/utilities.jl]: This second method overwrites the first method.
```julia
julia> using PkgDeps
[ Info: Precompiling PkgDeps [839e9fc8-855b-5b3c-a3b7-2833d3dd1f59]
WARNING: Method definition _get_pkg_name(Base.UUID) in module PkgDeps at /Users/nick/repos/PkgDeps.jl/src/utilities.jl:34 overwritten at /Users/nick/repos/PkgDeps.jl/src/utilities.jl:46.
  ** incremental compilation may be fatally broken for this module **
```

why not just one method that takes `registries`? `_get_pkg_name(uuid, reachable_registries(; kwargs...))` seems fine to me :) 

```suggestion
function _get_pkg_name(uuid::UUID, registries)
    for rego in registries
        for (pkg_name, pkg_entry) in rego.pkgs
            if pkg_entry.uuid == uuid
                return pkg_name
            end
        end
    end

    throw(NoUUIDMatch("No package found with the UUID $uuid"))
end
```

File: Project.toml [hunk 9]
```diff
@@ -5,12 +5,13 @@ version = "0.4.0"
...
 [deps]
 Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
+REPL = "3fa0cd96-eef1-5676-8a61-b3b8758bbffb"
 TOML = "fa267f1f-6049-4f14-aa54-33bafae1ed76"
 UUIDs = "cf7118a7-6976-5b1a-9a39-7adc72f591a4"
 
 [compat]
 TOML = "1.0"
-julia = "1"
+julia = "1.6"
```
Comment (nickrobinson251) [inline Project.toml]: why?

File: test/runtests.jl [hunk 10]
```diff
@@ -3,20 +3,20 @@ using Test
...
-const FOOBAR_REGISTRY = reachable_registries("Foobar"; depots=DEPOT)
+const FOOBAR_REGISTRY = first(reachable_registries("Foobar"; depots=DEPOT))
 
 
 @testset "internal functions" begin
     @testset "_get_pkg_name" begin
         @testset "uuid to name" begin
             expected = "Case1"
-            pkg_name = PkgDeps._get_pkg_name(UUID("00000000-1111-2222-3333-444444444444"); registries=[FOOBAR_REGISTRY])
+            pkg_name = PkgDeps._get_pkg_name(UUID("00000000-1111-2222-3333-444444444444"); depots=DEPOT)
```
Comment (nickrobinson251) [inline test/runtests.jl:14]: why this change?

Is this the fix mentioned?

> Fixed bug with users() would not allow you to look for usages of a package outside of pkg_registry. For example, if I wanted to look up all usages of DataFrames in an internal repository, I could not unless DataFrames was registered in the internal repository.

I think we could really do with a test for the "look up all usages of DataFrames in an internal repository" (something that failed before this PR but works in this one)

**Review by nickrobinson251 — COMMENTED** [review]
This PR is quite hard to review because it does a lot of different things that don't seem to be related, and doesn't add tests for the behaviour it's fixing, and especially because it both moves and edits code (making diffs hard to follow).

Would it be possible to break this into multiple PRs? 

I've tried to comment on the changes to the `users` behaviour (sorry about introducing a bug and thanks for working on fixing it!), but as i say, with the diffs / without a test i'm not sure i got it. 

I've not reviewed the "find alt packages" addition yet.

**Comment by mattBrzezinski** [comment]
> This PR is quite hard to review because it does a lot of different things that don't seem to be related, and doesn't add tests for the behaviour it's fixing, and especially because it both moves and edits code (making diffs hard to follow).
> 
> Would it be possible to break this into multiple PRs?
> 
> I've tried to comment on the changes to the `users` behaviour (sorry about introducing a bug and thanks for working on fixing it!), but as i say, with the diffs / without a test i'm not sure i got it.
> 
> I've not reviewed the "find alt packages" addition yet.

I will add in a test for the workflow which was failing, and try to explain it a bit more in-depth.

File: Project.toml (hunk 9 above)
Comment (mattBrzezinski) [reply Project.toml]: Oops, this should read `1.5, 1.6`.
I tested this w/ `1.3` and it doesn't work.

File: src/utilities.jl (hunk 2 above)
Comment (mattBrzezinski) [reply src/utilities.jl]: Neat didn't know `only` was a thing!

File: test/runtests.jl (hunk 10 above)
Comment (mattBrzezinski) [reply test/runtests.jl:14]: I did this in an attempt to simplify testing. The signature,

```julia
_get_pkg_name(::UUID; kwargs...)
```

Just wraps around

```julia
_get_pkg_name(::UUID, ::RegistryInstance[])
```

This cuts from 2 tests to 1 test

**Comment by mattBrzezinski** [comment]
Modified the test to handle the case where:

- `DownDep` is registered in `Foobar` registry
- We want to find all usages of `DownDep` in the `General` registry

File: Project.toml [hunk 11]
```diff
@@ -5,12 +5,13 @@ version = "0.4.0"
...
 [deps]
 Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
+REPL = "3fa0cd96-eef1-5676-8a61-b3b8758bbffb"
 TOML = "fa267f1f-6049-4f14-aa54-33bafae1ed76"
 UUIDs = "cf7118a7-6976-5b1a-9a39-7adc72f591a4"
 
 [compat]
 TOML = "1.0"
-julia = "1"
+julia = "1.5, 1.6"
```
Comment (mjram0s) [inline Project.toml]:
```suggestion
julia = "1.5"
```
This is the same

File: src/utilities.jl [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+"""
+Use levenshtein distance to find packages closely named to pkg_to_compare
+"""
+function _find_alternative_packages(pkg_to_compare::String, packages::Array)
+    pkg_to_compare = uppercase(pkg_to_compare)
+    pkg_distances = [(REPL.levenshtein(uppercase(pkg), pkg_to_compare), pkg) for pkg in packages]
+    sort!(pkg_distances)
+    counts = [count(x -> x[1] <= i, pkg_distances) for i in 0:2]
+    max_distance = max(0, searchsortedlast(counts, 8) - 1)
```
Comment (mjram0s) [inline src/utilities.jl]: Can you explain what's going on here? Might be helpful to define `8` as a variable or add a comment.

**Review by mjram0s — COMMENTED** [review]
I think you should test the functions in utilities.jl. I don't think the `_find_alternative_packages` functionality is tested here?

File: src/utilities.jl (hunk 12 above)
Comment (mattBrzezinski) [reply src/utilities.jl]: I simplified this down and left a comment for what the rules of the function are in the docstr

File: src/PkgDeps.jl [hunk 13]
```diff
@@ -132,8 +82,12 @@ Find the users of a given package.
 # Returns
 - `Array{String}`: All packages which are dependent on the given package.
 """
-function users(uuid::UUID; registries::Array{RegistryInstance}=reachable_registries())
-    pkg_name = _get_pkg_name(uuid; registries=registries)
+function users(
+    uuid::UUID;
+    registries::Array{RegistryInstance}=reachable_registries(),
+    depots::Union{String, Vector{String}}=Base.DEPOT_PATH
+)
```
Comment (mjram0s) [inline src/PkgDeps.jl:90]: Need to update docstring for this

**Review by mjram0s — APPROVED** [review]
